cpp/src/replace/replace.cu

```diff
@@ -13,31 +16,528 @@
+
+  }
+  template <typename col_type,
+            std::enable_if_t<not cudf::is_fixed_width<col_type>()>* = nullptr>
+  void operator()(cudf::column_view const& input,
+                  cudf::column_view const& replacement,
+                  cudf::mutable_column_view& output,
+                  cudaStream_t stream = 0) {
+    CUDF_FAIL("Non-fixed-width types are not supported for replace.");
+  }
+};
+
+
+/* --------------------------------------------------------------------------*/
+/**
+ * @brief Functor called by the `type_dispatcher` in order to invoke and instantiate
+ *        `replace_nulls` with the appropriate data types.
+ */
+/* ----------------------------------------------------------------------------*/
+struct replace_nulls_scalar_kernel_forwarder {
+  template <typename col_type,
+            std::enable_if_t<cudf::is_fixed_width<col_type>()>* = nullptr>
+  void operator()(cudf::column_view const& input,
+                  cudf::scalar const& replacement,
+                  cudf::mutable_column_view& output,
+                  cudaStream_t stream = 0)
+  {
+    cudf::size_type nrows = input.size();
+    cudf::experimental::detail::grid_1d grid{nrows, BLOCK_SIZE};
+
+    using ScalarType = cudf::experimental::scalar_type_t<col_type>;
+    auto s1 = static_cast<ScalarType const&>(replacement);
+
+    replace_nulls<<<grid.num_blocks, BLOCK_SIZE, 0, stream>>>(nrows,
+                                                              input.data<col_type>(),
+                                                              input.null_mask(),
+                                                              thrust::make_constant_iterator(s1.value()),
+                                                              output.data<col_type>());
+  }
+
+  template <typename col_type,
+            std::enable_if_t<not cudf::is_fixed_width<col_type>()>* = nullptr>
+  void operator()(cudf::column_view const& input,
+                  cudf::scalar const& replacement,
+                  cudf::mutable_column_view& output,
+                  cudaStream_t stream = 0) {
+    CUDF_FAIL("Non-fixed-width types are not supported for replace.");
+  }
+};
+
+} //end anonymous namespace
+
+
+namespace cudf {
+namespace detail {
+
+std::unique_ptr<cudf::column> replace_nulls(cudf::column_view const& input,
+                                            cudf::column_view const& replacement,
+                                            rmm::mr::device_memory_resource* mr,
+                                            cudaStream_t stream)
+{
+  if (input.size() == 0) {
+    return cudf::experimental::empty_like(input);
+  }
+
+  if (!input.has_nulls()) {
+    return std::make_unique<cudf::column>(input);
+  }
+
+  CUDF_EXPECTS(input.type() == replacement.type(), "Data type mismatch");
+  CUDF_EXPECTS(replacement.size() == input.size(), "Column size mismatch");
+  CUDF_EXPECTS(!replacement.has_nulls(), "Invalid replacement data");
```

Comment:
I'm not sure about this one. @harrism it seems to me like we should be able to replace a null value with another null value.